### PR TITLE
Add subscriber count checker before publishing costmap.

### PR DESCRIPTION
**Pull request by shivaang12**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #1028 |
| Primary OS tested on | Ubuntu) |
| Robotic platform tested on | gazebo |

---

## Description of contribution in a few bullet points
* Added `count_subscriber()` check before publishing costmap.

---

## Future work that may be required in bullet points
* I have locally tested it on my system but some unit and integration test might help.

### CODE REVIEW

File: nav2_costmap_2d/src/costmap_2d_publisher.cpp [hunk 1]
```diff
@@ -165,6 +165,13 @@ void Costmap2DPublisher::prepareCostmap()
 
 void Costmap2DPublisher::publishCostmap()
 {
+  if (node_->count_subscribers(costmap_raw_pub_->get_topic_name()) == 0) {
+    RCLCPP_DEBUG(node_->get_logger(), "costmap_raw subscriber not found!");
```
Comment (SteveMacenski) [inline nav2_costmap_2d/src/costmap_2d_publisher.cpp]: I'd remove the debugs -- also you're returning all costmap updates if no one's looking at the `raw` topic only.

These checks should occur in the method before each of their work is done but after a point that they're not independently blocking other costmap publish types

Comment (shivaang12) [reply nav2_costmap_2d/src/costmap_2d_publisher.cpp]: @SteveMacenski Is it okay if I do this check for each costmap or just `costmap_raw` is fine?

Comment (SteveMacenski) [reply nav2_costmap_2d/src/costmap_2d_publisher.cpp]: It needs to be each publisher

Comment (shivaang12) [reply nav2_costmap_2d/src/costmap_2d_publisher.cpp]: Is this look good?
```
void Costmap2DPublisher::publishCostmap()
{
  if (node_->count_subscribers(costmap_raw_pub_->get_topic_name()) > 0) {
    prepareCostmap();
    costmap_raw_pub_->publish(costmap_raw_);
  }
  float resolution = costmap_->getResolution();

  if (always_send_full_costmap_ || grid_.info.resolution != resolution ||
    grid_.info.width != costmap_->getSizeInCellsX() ||
    grid_.info.height != costmap_->getSizeInCellsY() ||
    saved_origin_x_ != costmap_->getOriginX() ||
    saved_origin_y_ != costmap_->getOriginY())
  {
    if (node_->count_subscribers(costmap_pub_->get_topic_name()) > 0) {
      prepareGrid();
      costmap_pub_->publish(grid_);
    }
  } else if (x0_ < xn_) {
    std::unique_lock<Costmap2D::mutex_t> lock(*(costmap_->getMutex()));
    // Publish Just an Update
    map_msgs::msg::OccupancyGridUpdate update;
    update.header.stamp = rclcpp::Time();
    update.header.frame_id = global_frame_;
    update.x = x0_;
    update.y = y0_;
    update.width = xn_ - x0_;
    update.height = yn_ - y0_;
    update.data.resize(update.width * update.height);

    if (node_->count_subscribers(costmap_update_pub_->get_topic_name()) > 0) {
      unsigned int i = 0;
      for (unsigned int y = y0_; y < yn_; y++) {
        for (unsigned int x = x0_; x < xn_; x++) {
          unsigned char cost = costmap_->getCost(x, y);
          update.data[i++] = cost_translation_table_[cost];
        }
      }
      costmap_update_pub_->publish(update);
    }
  }

  xn_ = yn_ = 0;
  x0_ = costmap_->getSizeInCellsX();
  y0_ = costmap_->getSizeInCellsY();
}
```

Comment (SteveMacenski) [reply nav2_costmap_2d/src/costmap_2d_publisher.cpp]: The update sub checker should happen before the mutex

**Pull request merged**